crates/bitwarden-uniffi/src/platform/fido2.rs

```diff
@@ -0,0 +1,246 @@
+
+        let mut platform = client.platform();
+        let mut fido2 = platform.fido2();
+        let ui = UniffiTraitBridge(self.1.as_ref());
+        let cs = UniffiTraitBridge(self.2.as_ref());
+        let mut auth = fido2.create_authenticator(&ui, &cs)?;
+
+        let result = auth.silently_discover_credentials(rp_id).await?;
+        Ok(result)
+    }
+}
+
+#[derive(uniffi::Object)]
+pub struct ClientFido2Client(pub(crate) ClientFido2Authenticator);
+
+#[uniffi::export]
+impl ClientFido2Client {
+    pub async fn register(
+        &self,
+        origin: String,
+        request: String,
+        client_data: ClientData,
+    ) -> Result<PublicKeyCredentialAuthenticatorAttestationResponse> {
+        let mut client = self.0 .0 .0.write().await;
+
+        let mut platform = client.platform();
+        let mut fido2 = platform.fido2();
+        let ui = UniffiTraitBridge(self.0 .1.as_ref());
+        let cs = UniffiTraitBridge(self.0 .2.as_ref());
+        let mut client = fido2.create_client(&ui, &cs)?;
+
+        let result = client.register(origin, request, client_data).await?;
+        Ok(result)
+    }
+
+    pub async fn authenticate(
+        &self,
+        origin: String,
+        request: String,
+        client_data: ClientData,
+    ) -> Result<PublicKeyCredentialAuthenticatorAssertionResponse> {
+        let mut client = self.0 .0 .0.write().await;
+
+        let mut platform = client.platform();
+        let mut fido2 = platform.fido2();
+        let ui = UniffiTraitBridge(self.0 .1.as_ref());
+        let cs = UniffiTraitBridge(self.0 .2.as_ref());
+        let mut client = fido2.create_client(&ui, &cs)?;
+
+        let result = client.authenticate(origin, request, client_data).await?;
+        Ok(result)
+    }
+}
+
+// Note that uniffi doesn't support external traits for now it seems, so we have to duplicate them
+// here.
+
+#[allow(dead_code)]
+#[derive(uniffi::Record)]
+pub struct CheckUserResult {
+    user_present: bool,
+    user_verified: bool,
+}
+
+#[uniffi::export(with_foreign)]
+#[async_trait::async_trait]
+pub trait UserInterface: Send + Sync {
+    async fn check_user(
+        &self,
+        options: CheckUserOptions,
+        credential: Option<CipherView>,
+    ) -> Result<CheckUserResult>;
+    async fn pick_credential_for_authentication(
+        &self,
+        available_credentials: Vec<Cipher>,
+    ) -> Result<CipherViewWrapper>;
+    async fn pick_credential_for_creation(
+        &self,
+        available_credentials: Vec<Cipher>,
+        new_credential: Fido2Credential,
+    ) -> Result<CipherViewWrapper>;
+}
+
+#[uniffi::export(with_foreign)]
+#[async_trait::async_trait]
+pub trait CredentialStore: Send + Sync {
+    async fn find_credentials(
+        &self,
+        ids: Option<Vec<Vec<u8>>>,
+        rip_id: String,
+    ) -> Result<Vec<Cipher>>;
+
+    async fn save_credential(&self, cred: Cipher) -> Result<()>;
+}
+
+struct UniffiTraitBridge<T>(T);
```

Comment:
The naming of this is a bit unfortunate, can we add some comments what purpose it serves?

Reply:
This is to work around the fact that we can't blanket implement the `bitwarden` trait over the copy in `bitwarden-uniffi`, I've added a comment now. I agree the name isn't great but it should be pretty limited in scope, let me know if you can think of a better name.